### PR TITLE
pings: RFC 197 (user accounting metrics)

### PR DESCRIPTION
**Pull request by ebrodymoore**
Add metrics listed in [RFC 197](https://docs.google.com/document/d/15fejG6sr68rlNPunVcJj9DU8O1SSVIQ9E8RYu5rIPU4/edit) to pings )all monthly):
- Number of created users
- Number of deleted users
- Number of resurrected users
- Number of retained users
- Number of churned users

### CODE REVIEW

**Comment by dadlerj** [comment]
:tada: thanks @ebrodymoore  

@tsenart before we merge new `event_logs` analysis for pings, how should we validate/test performance characteristics? Is this something you can peek at and get a sense for risk? Just run this with some fake data on the order of sg.com's count of rows in `event_logs` to see how it looks? 

I can see a few small optimizations (e.g. grouping the `event_logs` table on `user_id` before joining with `users` maybe), but I'm not even sure if such things are necessary. It'd be great to get this into 3.18 if possible.

**Comment by tsenart** [comment]
> @tsenart before we merge new event_logs analysis for pings, how should we validate/test performance characteristics? Is this something you can peek at and get a sense for risk? Just run this with some fake data on the order of sg.com's count of rows in event_logs to see how it looks?

I'd just run the query against the sourcegraph.com database via psql or another client and look at its performance first (i.e. total run-time). If it's slow then we could EXPLAIN ANALYZE it with https://flame-explain.com/

**Comment by ebrodymoore** [comment]
> If it's slow then we could EXPLAIN ANALYZE it with https://flame-explain.com/

What would you characterize as slow?

**Comment by tsenart** [comment]
Anything that would go beyond our timeout and would break pings. I think that's 60s but not sure.

File: cmd/frontend/internal/app/pkg/updatecheck/client.go [hunk 1]
```diff
@@ -280,6 +280,8 @@ func updateBody(ctx context.Context) (io.Reader, error) {
 		return nil, err
 	}
 
+	fmt.Println(string(contents))
```
Comment (ebrodymoore) [inline cmd/frontend/internal/app/pkg/updatecheck/client.go:283]: @dadlerj I just tested this again and it seems ready to go - to merge, we need to get rid of this, correct? 

Could you approve the PR (if it looks good to you, obviously)?

Comment (tsenart) [reply cmd/frontend/internal/app/pkg/updatecheck/client.go:283]: We should probably remove this`fmt.Println`?

File: cmd/frontend/types/types.go [hunk 2]
```diff
@@ -380,6 +380,13 @@ type Event struct {
 	Timestamp       time.Time
 }
 
+type GrowthStatistics struct {
+	DeletedUsers 		int32
+	CreatedUsers 		int32
+	ResurrectedUsers 	int32
+	ChurnedUsers		int32
+	RetainedUsers 		int32
+
```
Comment (tsenart) [inline cmd/frontend/types/types.go:389]: There seems to be a missing closing bracket `}`. Was this compiling 😕 ?

File: cmd/frontend/types/types.go [hunk 3]
```diff
@@ -380,6 +380,13 @@ type Event struct {
 	Timestamp       time.Time
 }
 
+type GrowthStatistics struct {
```
Comment (tsenart) [inline cmd/frontend/types/types.go:383]: Please add a Go doc comment to this type (for instance, mentioning the time period being a month).

File: cmd/frontend/internal/usagestats/growth.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+import (
+	"context"
+
+	"github.com/sourcegraph/sourcegraph/cmd/frontend/types"
+	"github.com/sourcegraph/sourcegraph/internal/db/dbconn"
+)
+
+func GetGrowthStatistics(ctx context.Context) (*types.GrowthStatistics, error) {
+	const q = `
```
Comment (tsenart) [inline cmd/frontend/internal/usagestats/growth.go:13]: Please add a comment to the query  at the top like `// source: cmd/frontend/internal/usagestats/growth.go:GetGrowthStatistics`. It helps us identify where a query comes from easily when debugging slow queries in production.

File: cmd/frontend/internal/usagestats/growth.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+      OR deleted_month IS NULL)) AS churned_users,
+  COUNT(*) FILTER (
+  WHERE
+    current_month = TRUE
+    AND previous_month = TRUE
+    AND created_month < DATE_TRUNC('month', now())
+    AND (deleted_month < DATE_TRUNC('month', now())
+      OR deleted_month IS NULL)) AS retained_users
+FROM
+  sub
```
Comment (tsenart) [inline cmd/frontend/internal/usagestats/growth.go:76]: How did the testing in the production database go? Can you share your findings here for future reference?

Comment (ebrodymoore) [reply cmd/frontend/internal/usagestats/growth.go:76]: Yes - Dan helped me and we took out a join or two and the time dropped to a couple seconds so not worried about performance issues with this specific query!

File: cmd/frontend/types/types.go (hunk 2 above)
Comment (ebrodymoore) [reply cmd/frontend/types/types.go:392]: It was 🤔 - I added the closing bracket and it compiled so let's go with that!

File: cmd/frontend/internal/app/pkg/updatecheck/handler_test.go [hunk 6]
```diff
@@ -220,6 +222,7 @@ func TestSerializeFromQuery(t *testing.T) {
 		"automation_usage": null,
 		"code_intel_usage": null,
 		"search_usage": null,
+		"growth_statistics": null
```
Comment (uwedeportivo) [inline cmd/frontend/internal/app/pkg/updatecheck/handler_test.go:225]: you're missing commas

**Pull request merged**